osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs

```diff
@@ -166,5 +151,89 @@ private void attemptStart()
             if (Preview?.Start() != true)
                 playing.Value = false;
         }
+
+        private class Button : CompositeDrawable
+        {
+            private readonly SpriteIcon icon;
+            private readonly LoadingSpinner loadingSpinner;
+
+            private readonly IBindable<bool> playing;
+            private readonly IBindable<bool> disabled;
+
+            [Resolved]
+            private OsuColour colours { get; set; }
+
+            public bool Loading
+            {
+                set
+                {
+                    if (value)
+                    {
+                        icon.FadeTo(0.5f, transition_duration, Easing.OutQuint);
+                        loadingSpinner.Show();
+                    }
+                    else
+                    {
+                        icon.FadeTo(1, transition_duration, Easing.OutQuint);
+                        loadingSpinner.Hide();
+                    }
+                }
+            }
+
+            public Button(PlayButton player)
+            {
+                playing = player.Playing.GetBoundCopy();
+                disabled = player.Disabled.GetBoundCopy();
+
+                InternalChildren = new Drawable[]
+                {
+                    icon = new SpriteIcon
+                    {
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        FillMode = FillMode.Fit,
+                        RelativeSizeAxes = Axes.Both,
+                        Icon = FontAwesome.Solid.Play,
+                    },
+                    loadingSpinner = new LoadingSpinner
+                    {
+                        Size = new Vector2(15),
+                    },
+                };
+            }
+
+            protected override void LoadComplete()
+            {
+                base.LoadComplete();
+
+                playing.BindValueChanged(_ => updateDisplay());
+                disabled.BindValueChanged(_ => updateDisplay());
+                updateDisplay();
+            }
+
+            protected override bool OnHover(HoverEvent e)
+            {
+                updateDisplay();
+                return base.OnHover(e);
+            }
+
+            protected override void OnHoverLost(HoverLostEvent e)
+            {
+                updateDisplay();
+                base.OnHoverLost(e);
+            }
+
+            private void updateDisplay()
```

Comment:
I'd probably make this a `virtual protected` to avoid duplicating all the calling logic of this function (it's being bound in three different events). Make sure to attach xmldoc mentioning it will be called on hover or disabled state changes.

In fact, if you send in hovered and disabled as argument you may be able to keep the disabled bindable private.

Reply:
Well, one problem is that hovering over the icon button is different from hovering over `HeaderPlayButton`, so the `virtual protected` method would have to accept hovering state of both the `PlayButton` and the icon button, which sounds kinda too much from just seperating them.